assets/js/wc-gateway-ppec-generate-cart.js

```diff
@@ -30,52 +30,67 @@
 			$( '#woo_pp_ec_button_product > *' ).css( 'pointer-events', 'none' );
 		} );
 
+	var variation_valid = true;
+	var fields_valid = true;
+	var update_button = function() {
+		$( '#woo_pp_ec_button_product' ).trigger( ( variation_valid && fields_valid ) ? 'enable' : 'disable' );
+	};
+
+	var validate_form = function() {
+		silent_validation = true;
+		fields_valid = true;
+		form.find( 'select:enabled, input:enabled, textarea:enabled' ).each( function() {
+			fields_valid = fields_valid && this.checkValidity();
+		} );
+		silent_validation = false;
+		update_button();
+	};
+
 	// It's a variations form, button availability should depend on its events
 	if ( $( '.variations_form' ).length ) {
-		$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+		variation_valid = false;
 
 		$( '.variations_form' )
 		.on( 'show_variation', function( event, form, purchasable ) {
-			$( '#woo_pp_ec_button_product' ).trigger( purchasable ? 'enable' : 'disable' );
+			variation_valid = purchasable;
+			update_button();
 		} )
 		.on( 'hide_variation', function() {
-			$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+			variation_valid = false;
+			update_button();
 		} );
 	}
 
-	var get_attributes = function() {
-		var select = $( '.variations_form' ).find( '.variations select' ),
-			data   = {},
-			count  = 0,
-			chosen = 0;
-
-		select.each( function() {
-			var attribute_name = $( this ).data( 'attribute_name' ) || $( this ).attr( 'name' );
-			var value	  = $( this ).val() || '';
-
-			if ( value.length > 0 ) {
-				chosen++;
-			}
-
-			count++;
-			data[ attribute_name ] = value;
-		} );
-
-		return {
-			'count'      : count,
-			'chosenCount': chosen,
-			'data'       : data
-		};
-	};
+	// Disable the button if there are invalid fields in the product page (like required fields from Product Addons)
+	var silent_validation;
+	var form = $( 'form.cart' );
+	form.get( 0 ).addEventListener( 'invalid', function( e ) {
```

Comment:
I'm not sure I understand why this event listener is added. Can you explain?

Reply:
This listener blocks the `oninvalid` event from reaching "the browser". Without it, an annoying "This field is required" pop-over would appear over the invalid fields every time another field changes. That's why I called the flag `silent_validation`.

Reply:
Adding to that, an `oninvalid` event is fired on a field when:
- The user clicks the "Submit" button and the field fails validation.
- `inputElement.checkValidity()` is called, and the field fails validation.

Reply:
Ah, I see what you mean - since we're using `checkValidity` and it can fire an `invalid` event, we should listen to it and cancel the HTML 5 validation that can be displayed.

I did a bit of digging, and from what I can see the `invalid` event is not what prompts the browser to display the HTML 5 validation errors. I also tested this by removing the `e.preventDefault` line, and I see that the validation errors are not displayed when I change any of the form fields. This works on all browsers (didn't personally check IE11).

So, it looks like `e.preventDefault` is no needed for `invalid` events. The form will display errors only when it's submitted.

Is there another issue I'm missing?

Reply:
Hah! Good catch. My first try was to use `reportValidity()`, and that **did** prompt the HTML5 validation pop-overs. That's not needed with `checkValidity()`. I ❤️ these APIs that are almost the same but different in sutble ways...

I've removed the whole listener in https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/pull/507/commits/b61ec7d1fa7d09c346fd5391f400c4280002992e

Reply:
I know that feeling... :)